mypyc: Fix C string encoding

saleemrashid opened the pull request:
Using `repr()` to encode C strings relied on implementation-defined behaviour and had a number of bugs, for example:

* The C string could inadvertently contain trigraphs.

* Valid hexdecimal characters following a hexadecimal escape sequence would be parsed as part of the escape sequence.

Octal escape sequences are used for unprintable characters because they do not have the same issue as hexadecimal escape sequences.

It would be possible to mitigate the issue with hexadecimal escape sequences by emitting multiple string literals. However, the complexity is not worth it, especially as the output is not designed to be human-readable.

As an ad hoc test, I generated a Python module containing string literals for every possible character:

```python
with open("literals.py", "w") as f:
    f.write(
        "LITERALS = [{}]".format(
            ", ".join([repr(bytes([x, ord("0")])) for x in range(256)])
        )
    )
```

Then I compiled the module with Mypyc and checked it produced the correct byte strings:

```python
import literals

assert literals.__file__.endswith(".so")

for x in range(256):
    assert literals.LITERALS[x] == bytes([x, ord("0")])
```

msullivan commented on mypyc/emitmodule.py:
```diff
@@ -421,8 +449,7 @@ def encode_as_c_string(s: str) -> Tuple[str, int]:
     """Produce a single-escaped, quoted C string and its size from a bytes"""
-    # This is a kind of abusive way to do this...
-    escaped = repr(b)[2:-1].replace('"', '\\"')
+    escaped = ''.join(map(C_CHAR_MAP.__getitem__, b))
```
I'd write this as a list comprehension instead of a map

msullivan commented on mypyc/emitmodule.py:
```diff
@@ -67,6 +68,33 @@
 # A list of (file name, file contents) pairs.
 FileContents = List[Tuple[str, str]]
 
+# The C standard specifies that any number of valid hexadecimal characters are
```
I think it's worth moving this out of `emitmodule` at this point. `mypyc.cstring`, maybe?

msullivan replied:
(And then encode_as_c_string and encode_bytes_as_c_string too)

msullivan commented on mypyc/emitmodule.py:
```diff
@@ -67,6 +68,33 @@
+# These assignments must be done after string.printable because they are
+# overrides for the printable characters that need to be escaped in string
+# literals.
+C_CHAR_MAP[ord('\'')] = '\\\''
```
These are all generated the same way (adding a backslash), so I think it would be an improvement to generate them all mechanically from a list of characters to escape.

saleemrashid replied:
I considered this, but the neatest way I can think to do this is:

```python
for x in ("'", '"', '\\', 'a', 'b', 'f', 'n', 'r', 't', 'v'):
    escaped = '\\{}'.format(x)
    value = escaped.encode('ascii').decode('unicode_escape')
    C_CHAR_MAP[ord(value)] = escaped
```

I am not sure whether this is an improvement. What do you think?

saleemrashid commented:
I've force-pushed some changes, i.e. rewrote the comments, and refactored the escaped strings to raw string literals.

I will try and implement your suggestions shortly.

> I'm also sad about the octal thing but oh well.

@msullivan My original prototype used `itertools.groupby` to group blocks of printable characters and unprintable characters so you could safely use hexadecimal escape sequences. It worked fine but it seemed like unnecessary complexity.

msullivan commented:
> @msullivan My original prototype used `itertools.groupby` to group blocks of printable characters and unprintable characters so you could safely use hexadecimal escape sequences. It worked fine but it seemed like unnecessary complexity.

I agree that using octal is preferable to writing complex code that avoids it.

saleemrashid commented on mypyc/cstring.py:
```diff
@@ -0,0 +1,53 @@
+
+# These assignments must come last because we prioritize the escape
+# sequences in the C standard over any other representation.
+CHAR_MAP[ord('\'')] = r'\''
```
@msullivan This is the simplest method I have been able to devise to generate these escape sequences:

```python
for x in ("'", '"', '\\', 'a', 'b', 'f', 'n', 'r', 't', 'v'):
    escaped = '\\{}'.format(x)
    value = escaped.encode('ascii').decode('unicode_escape')
    C_CHAR_MAP[ord(value)] = escaped
```

While I am not a huge fan of manually listing each escape sequence, I dislike this method more.

msullivan replied:
The encode/decode is a little hokey but I think I like this more

saleemrashid replied:
The main reason I dislike this is that we're calling into Python's Unicode APIs. There's also `codecs.escape_decode` that handles ASCII only, but it's undocumented.

msullivan commented:
Could you also remove `-Wno-trigraph` from our C flags in `build.py` <_<

saleemrashid commented on mypyc/cstring.py:
```diff
@@ -0,0 +1,53 @@
+    return encode_bytes_as_c_string(s.encode('utf-8'))
+
+
+def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
```
This is what an implementation with hexadecimal escape sequences, instead of octal, would look like:

```diff
diff --git i/mypyc/cstring.py w/mypyc/cstring.py
index 37f3ab38..0557a216 100644
--- i/mypyc/cstring.py
+++ w/mypyc/cstring.py
@@ -19,9 +19,11 @@ octal digits.
 """
 
 import string
+import itertools
 from typing import Tuple
 
-CHAR_MAP = ['\\{:03o}'.format(i) for i in range(256)]
+HEX_DIGITS = frozenset(string.hexdigits)
+CHAR_MAP = {}
 
 # It is safe to use string.printable as it always uses the C locale.
 for c in string.printable:
@@ -49,5 +51,13 @@ def encode_as_c_string(s: str) -> Tuple[str, int]:
 
 def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
     """Produce a quoted C string literal and its size, for a byte string."""
-    escaped = ''.join([CHAR_MAP[i] for i in b])
+    escaped = ''
+    for printable, group in itertools.groupby(b, key=CHAR_MAP.__contains__):
+        if printable:
+            s = ''.join([CHAR_MAP[i] for i in group])
+            if len(escaped) and s[0] in HEX_DIGITS:
+                escaped += '" "'
+            escaped += s
+        else:
+            escaped += ''.join(['\\x{:02X}'.format(i) for i in group])
     return '"{}"'.format(escaped), len(b)
```

msullivan commented:
Thanks!
